### PR TITLE
[ISSUE#23]PUT articles/{id}- add tests

### LINKED ISSUE #23: PUT /articles/{id} - tests
Tests for PUT /articles/{id} endpoint

### PR DESCRIPTION
**Pull request by Justyna-KO**
### Scope of changes:
Tests PUT articles/{id} for responses:
- 200: article updated
- 201: article created
- 400: Bad request - malformed JSON
- 404: Unauthorized
- -422: Invalid article supplied

## Any other comments?
The test for the 404 response is missing, any ideas on how to get 404 in response for this endpoint? 

Before submitting Pull Request, I've ensured that:

- [x] The PR is associated with the relevant Issue before its creation.
- [x] All new and existing tests passed
- [x] My commit messages follow the `kawqa-gad-playwright` project's [git commit message format](https://github.com/kat-kan/kawqa-gad-playwright/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: tests/api/articles/articles-is-put.spec.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+test.describe('PUT articles/{id} endpoint tests', async () => {
+  const articles = `/api/articles`;
+  const newTitle =
+    'How to start writing API tests? ' +
+    (Math.floor(Math.random() * (1000 - 100 + 1)) + 100);
+  const newContent = 'Join KawQA z automatu';
+  const articleDate: string = '2024-07-14T15:44:31Z';
+  const articleImage: string =
+    'src\\test-data\\images\\Roasted_coffee_beans.jpg';
+  const articleId: number = Math.floor(Math.random() * (1000 - 100 + 1)) + 100;
```
Comment (kat-kan) [inline tests/api/articles/articles-is-put.spec.ts:15]: Looks like this expression is the very same as in line 10 so it could be saved in the variable. 
Another thing is that according to GAD repo page, validation of unique title is currently (from GAD 2.7.0) triggered with feature flag. So unless feature flag is on, there is no need to generate unique title

File: tests/api/articles/articles-is-put.spec.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      .soft(responseBody.user_id.toString())
+      .toEqual(testUsers.regularUser.id.toString());
+    expect.soft(responseBody.title).toBe(newTitle);
+    expect.soft(responseBody.body).toBe(newContent);
+    expect.soft(responseBody.date).toBe(articleDate);
+    expect.soft(responseBody.image).toBe(articleImage);
+    expect.soft(typeof responseBody.id === 'number').toBe(true);
+  });
+
+  test('Returns 201 OK status code when updating article', async ({
```
Comment (kat-kan) [inline tests/api/articles/articles-is-put.spec.ts:50]: This is a specific situation when PUT is used to create new article, I think better description is "...when creating article" or "...when creating article using PUT"

File: tests/api/articles/articles-is-put.spec.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        body: newContent,
+        date: articleDate,
+        image: articleImage,
+      },
+    });
+    // Then
+    expect.soft(response.status()).toBe(HttpStatusCode.Unauthorized);
+  });
+
+  test('Returns 422 Unprocessable content after updating  article with wrong date format', async ({
```
Comment (kat-kan) [inline tests/api/articles/articles-is-put.spec.ts:121]: looks like there is an additional space between updating and article

**Review by kat-kan — COMMENTED** [review]
This is my initial review - I do not have access to coding env this week so I'll get back to this PR

**Comment by Slawomir-DKl** [comment]
@Justyna-KO The latest GAD version (2.7.2) brings another logic which is not yet tested:

_using PUT with empty body will return error 422 (comments, users, articles)_ 

If you think about developing the tests further, it would be quite nice extension

File: tests/api/articles/article-id-put.spec.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  }) => {
+    const response: APIResponse = await request.put(articles, {
+      headers: setHeaders,
+      data: {},
+    });
+
+    expect(response.status()).toBe(HttpStatusCode.UnprocessableEntity);
+  });
+
+  test.describe('PUT articles/{id} endpoint tests with enabled feature', async () => {
```
Comment (kat-kan) [inline tests/api/articles/article-id-put.spec.ts:143]: I like the additional tests, good job 👍🏼 
I do have small concern about naming - imagine that you do not have access to test code and need to understand test results based on test report only. So based on the test description we know that some feature was enabled (or that validation on title was enabled) but it's still not clear.
Also - if we add test.describe() for group of tests, report will show this + value from each test. Which means that we can write about this enabled validation only on test group level, it will be enough.
To sum up, I'd recommend to try adding more specific info about enabled validation of unique title, removing it from test() level. Lemme know what You think and if such report would be more readable :)

File: tests/api/articles/article-id-put.spec.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        .soft(responseBody.user_id.toString())
+        .toEqual(testUsers.regularUser.id.toString());
+      expect.soft(responseBody.title).toBe(`${newTitle}${randomNumber}`);
+      expect.soft(responseBody.body).toBe(newContent);
+      expect.soft(responseBody.date).toBe(articleDate);
+      expect.soft(responseBody.image).toBe(articleImage);
+      expect.soft(typeof responseBody.id === 'number').toBe(true);
+    });
+
+    test('Returns 201 Created status code when updating article - enabled validation on the title', async ({
```
Comment (kat-kan) [inline tests/api/articles/article-id-put.spec.ts:180]: Aren't these 2 descriptions (this one and the one above) mixed? 201 Created should be returned when sth is..created?

Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts]: Yes, you're right.

File: tests/api/articles/article-id-put.spec.ts (hunk 4 above)
Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts]: good point, I changed it

File: tests/api/articles/article-id-put.spec.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    expect.soft(responseBody.image).toBe(articleImage);
+    expect.soft(typeof responseBody.id === 'number').toBe(true);
+  });
+
+  test('Returns 400 Bad Request after sending malformed JSON', async ({
+    request,
+  }) => {
+    const malformedJson: string = `{
+      "user_id": "${testUsers.regularUser.id}",
+      "title: ${newTitle},  // error: missing closing quotation mark
```
Comment (Slawomir-DKl) [inline tests/api/articles/article-id-put.spec.ts:84]: The problem is not only in the closing quotation mark - the whole string is formatted incorrectly.

I tried to construct proper JSON (nice exercise, by the way).
If I properly understand the [Playwright documentation](https://playwright.dev/docs/api/class-apirequestcontext#api-request-context-put), `data` option serializes its parameter to JSON, therefore I used JSON.parse method to debug - I added two lines for debugging purposes only: 
```
const malJson = JSON.parse(malformedJson);
console.log(malJson);
```
This way I stated that here the problems are:
- newline characters (parser doesn't recognize properties, as they are after newlines)
- lacking closing quotation mark, as you stated
- the string in the image should have groups of four slashes instead of two slashes (as it is converted twice)

So, to get a different response status (which was my goal as I checked the test negatively), I had to use the one-line const:
```const malformedJson: string = `{"user_id": "${testUsers.regularUser.id}", "title": "${newTitle}", "body": "${newContent}", "date": "${articleDate}", "image": "src\\\\test-data\\\\images\\\\Roasted_coffee_beans.jpg"}`;```
Such string gives proper status 422 (as the title is not unique)

By the way, I have to check other malformed JSONs in our project - they are constructed similarly.

Of course, the goal is achieved - your string is finally not a proper JSON and the response is 400, but I think we should check each test negatively, too, to understand how it works

If you want to use my one-liner, it should be enough to remove one quotation mark from the string to get the 400 status

File: tests/api/articles/article-id-put.spec.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  }) => {
+    const response: APIResponse = await request.put(articles, {
+      headers: setHeaders,
+      data: {},
+    });
+
+    expect(response.status()).toBe(HttpStatusCode.UnprocessableEntity);
+  });
+
+  test.describe('PUT articles/{id} endpoint tests with enabled feature_validate_article_title', async () => {
```
Comment (Slawomir-DKl) [inline tests/api/articles/article-id-put.spec.ts:143]: Great work with the tests with the enabled feature. I learnt a lot from your code. Nevertheless, I failed in attempts to switch the validation on for the first test and switch it off for the second test, so maybe the flag doesn't impact PUT requests. It would be great to create a negative test to show that the flag works and the title is not validated when the flag is set to false. 
Now when I changed the title to "a" in the second test, I got 422 each time (with response body `{ error: { message: 'Field "title" is not unique!' } }`), regardless of setting the flag `true` or `false`.

**Review by Slawomir-DKl — COMMENTED** [review]
Please take a look at my comments. I'm not sure if they require changes - I leave it to you.

File: tests/api/articles/article-id-put.spec.ts (hunk 6 above)
Comment (kat-kan) [reply tests/api/articles/article-id-put.spec.ts:84]: I need to admit that I do not quite get it without playing around with all these examples and yes, this malformed json is used in all tests. Can you please talk about it during Monday meeting?

File: tests/api/articles/article-id-put.spec.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+import { APIResponse, expect, test } from '@playwright/test';
+
+test.describe('PUT articles/{id} endpoint tests', async () => {
+  const articles = `/api/articles`;
+  const randomNumber = Math.floor(Math.random() * (1000 - 100 + 1)) + 100;
+  const newTitle = 'How to start writing API tests? ';
+  const newTitle2 = 'How to start writing API tests?' + randomNumber;
+  const newTitle3 = 'How to start writing good API tests?' + randomNumber;
+  const newContent = 'Join KawQA z automatu';
+  const articleDate: string = '2024-07-14T15:44:31Z';
```
[the remark this answers is not in the document]
Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts:13]: good point, I've changed it.

File: tests/api/articles/article-id-put.spec.ts (hunk 6 above)
Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts:87]: we will discuss it at the meeting

File: tests/api/articles/article-id-put.spec.ts (hunk 7 above)
Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts:172]: I have added corrections, please review.
I've  also added a negative test.

File: tests/api/articles/article-id-put.spec.ts [hunk 9]
```diff
@@ -2,18 +2,21 @@ import { HttpStatusCode } from '@_src_api/enums/api-status-code.enum';
...
+import { customDate } from 'test-data/shared/date.generator';
 
 test.describe('PUT articles/{id} endpoint tests', async () => {
   const articles = `/api/articles`;
   const randomNumber = Math.floor(Math.random() * (1000 - 100 + 1)) + 100;
-  const newTitle = 'How to start writing API tests? ';
+  const newTitle = 'How to start writing API tests?';
   const newTitle2 = 'How to start writing API tests?' + randomNumber;
   const newTitle3 = 'How to start writing good API tests?' + randomNumber;
+  const oldTilte = 'The difference between Scrum and Kanban';
```
Comment (Slawomir-DKl) [inline tests/api/articles/article-id-put.spec.ts:13]: I've just found that this is a typo. Now: `oldTilte` but it should be: `oldTitle` (replace in the whole file)

File: tests/api/articles/article-id-put.spec.ts (hunk 7 above)
Comment (Slawomir-DKl) [reply tests/api/articles/article-id-put.spec.ts:172]: @Justyna-KO Your new tests allowed me to understand that this feature flag works in PUT requests :-).

What do you think about renaming the new tests (both 200 and 422)? Now the name states `when updating article with the same title`. This can be misleading - firstly, I understood that the title has to be the same as the previous title of the same article, while the title has to be the same as any title in the DB (i.e. it has to be non-unique). What about `when updating the article with the title equal to another article title`?

**Review by Slawomir-DKl — CHANGES_REQUESTED** [review]
Thank you for the changes. I propose two small changes.

File: tests/api/articles/article-id-put.spec.ts (hunk 7 above)
Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts:172]: ok, this will be clearer

File: tests/api/articles/article-id-put.spec.ts (hunk 9 above)
Comment (Justyna-KO) [reply tests/api/articles/article-id-put.spec.ts:13]: I've corrected it.

**Review by Slawomir-DKl — DISMISSED** [review]
Thank you for the changes. I think the tests are ready to merge

**Review by Slawomir-DKl — APPROVED** [review]
